pkg/bpf/bpffs.go

```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+	return num > 1, nil
+}
+
+// isBPFFS checks whether the current mapRoot:
+// - has any mount
+// - if yes, it checks whether the filesystem of that mount is BPFFS
+func isBPFFS() (bool, bool, error) {
+	// Check whether mapRoot has any mount (by checking /proc/mounts)
+	scanner, err := newMountsScanner()
+	if err != nil {
+		return false, false, err
+	}
+
+	hasMount := false
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), mapRoot) {
+			hasMount = true
+			break
 		}
+	}
+	scanner.Close()
 
+	if !hasMount {
+		return false, false, nil
 	}
 
+	// Check whether the mount filesystem is BPFFS
 	var fsdata unix.Statfs_t
 	if err := unix.Statfs(mapRoot, &fsdata); err != nil {
-		return fmt.Errorf("BPF filesystem path %s is not mounted", mapRoot)
+		log.WithError(err).Errorf("unable to get information about %s mount", mapRoot)
+		return true, false, err
 	}
 
 	// This is the value of the BPF Filesystem defined in
 	// uapi/linux/magic.h The magic value can potentially be misleading if
 	// the BPF filesystem is mounted in the host and then volume mapped
 	// into a container.
 	magic := uint32(0xCAFE4A11)
-	if uint32(fsdata.Type) != magic {
-		log.WithField(logfields.Path, mapRoot).Warningf("BPF root is not a BPF filesystem (%#x != %#x)",
-			uint32(fsdata.Type), magic)
+
+	if uint32(fsdata.Type) == magic {
+		return true, true, nil
+	}
+	return true, false, nil
+}
+
+// checkOrMountCustomLocation tries to check or mount the BPF filesystem in the
+// given path.
+func checkOrMountCustomLocation(bpfRoot string) error {
+	SetMapRoot(bpfRoot)
+
+	// Check whether the custom location has a BPFFS mount.
+	mounted, bpffsInstance, err := isBPFFS()
+	if err != nil {
+		return err
+	}
+
+	// If the custom location has no mount, let's mount BPFFS there.
+	if !mounted {
+		SetMapRoot(bpfRoot)
+		if err := mountFS(); err != nil {
+			return err
+		}
+	}
+
+	// If the custom location already has a mount with some other filesystem than
+	// BPFFS, return the error.
+	if !bpffsInstance {
+		return fmt.Errorf("mount in the custom directory %s has a different filesystem than BPFFS", bpfRoot)
+	}
+
+	return nil
+}
+
+// checkOrMountCustomLocation tries to check or mount the BPF filesystem in
+// standard locations, which are:
+// - /sys/fs/bpf
+// - /run/cilium/bpffs
+// There is a procedure of determining which directory is going to be used:
+// 1. Checking whether BPFFS filesystem is mounted in /sys/fs/bpf.
+// 2. If there is no mount, then mount BPFFS in /sys/fs/bpf and finish there.
+// 3. If there is a BPFFS mount, finish there.
+// 4. If there is a mount, but with the other filesystem, then it means that most
+//    probably Cilium is running inside container which has mounted /sys/fs/bpf
+//    from host, but host doesn't have proper BPFFS mount, so that mount is just
+//    the empty directory. In that case, mount BPFFS under /run/cilium/bpffs.
+func checkOrMountDefaultLocations() error {
```

Comment:
The function description is not the same as the function's name

Reply:
done